### PR TITLE
util: Adjust binary collate behavior to do pattern matching in bytes

### LINKED ISSUE #50393: Join between blob type with matching returns incorrect result.
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
create table t2(a blob);
create table t3(a blob);
insert into t2 values(0xC2A0);
insert into t3 values(0xC2);
select * from t2,t3 where t2.a like concat("%",t3.a,"%");
```

### 2. What did you expect to see? (Required)

Returns one row, as MySQL(8.0.33) does.

```sql
mysql> select * from t2,t3 where t2.a like concat("%",t3.a,"%");
+------------+------------+
| a          | a          |
+------------+------------+
| 0xC2A0     | 0xC2       |
+------------+------------+
1 row in set (0.01 sec)
```

### 3. What did you see instead (Required)

Returns empty set.

```sql
> select * from t2,t3 where t2.a like concat("%",t3.a,"%");
Empty set (0.00 sec)
```

It looks weird since in the case below, TiDB works well.
```sql
create table t2(a blob);
create table t3(a blob);
insert into t2 values(0xC2A020);
insert into t3 values(0xC2A0);
select * from t2,t3 where t2.a like concat("%",t3.a,"%");
```

In this case, TiDB returns one row:
```sql
> select * from t2,t3 where t2.a like concat("%",t3.a,"%");
+------+------+
| a    | a    |
+------+------+
|      |      |
+------+------+
1 row in set (0.00 sec)
```

which is similar with MySQL.(The difference in display is reported in #50392).

```sql
mysql> select * from t2,t3 where t2.a like concat("%",t3.a,"%");
+------------+------------+
| a          | a          |
+------------+------------+
| 0xC2A020   | 0xC2A0     |
+------------+------------+
1 row in set (0.02 sec)
```

The charsets of TiDB and MySQL are the same(utf8mb4).

### 4. What is your TiDB version? (Required)

```sql
>  SELECT tidb_version();
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()

       |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v7.4.0
Edition: Community
Git Commit Hash: 38cb4f3312be9199a983c0ef282d2ea2e28a7824
Git Branch: heads/refs/tags/v7.4.0
UTC Build Time: 2023-10-10 14:18:50
GoVersion: go1.21.1
Race Enabled: false
Check Table Before Drop: false
Store: tikv |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

### PR DESCRIPTION
**Pull request by yibin87**
### What problem does this PR solve?

Issue Number: close #50393

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/util/collate/bin.go [hunk 1]
```diff
@@ -65,6 +65,16 @@ func (*binPaddingCollator) Pattern() WildcardPattern {
 	return &binPattern{}
 }
 
+// binPureCollator match pattern in bytes
+type binPureCollator struct {
```
Comment (YangKeao) [inline pkg/util/collate/bin.go:69]: How about modify the behavior of `binPattern` directly? I don't think there is anything depending on the original wrong behavior 🤔. Is there any special reason to keep the original `binPattern`?

File: pkg/util/stringutil/string_util.go [hunk 2]
```diff
@@ -228,53 +281,76 @@ func CompileLike2Regexp(str string) string {
 	return string(result)
 }
 
-// DoMatchBytes is a adapter for `DoMatchInner`, `str` can only be an ascii string.
+// DoMatchBytes is an adapter for `DoMatchInner`, `str` can only be an ascii string.
 func DoMatchBytes(str string, patChars, patTypes []byte) bool {
```
Comment (YangKeao) [inline pkg/util/stringutil/string_util.go:285]: Also, I think we don't need to distinguish `DoMatchBytes` and `DoMatchPureBytes`. We can directly modify the behavior of `DoMatchBytes` (under proper tests).

File: pkg/util/stringutil/string_util.go [hunk 3]
```diff
@@ -228,53 +281,76 @@ func CompileLike2Regexp(str string) string {
...
+	bytes := []byte(str)
+	lenBytes := len(bytes)
+	lenPatWeights := len(patChars)
+	return DoMatchInner(lenPatWeights, lenBytes, patTypes, func(a, b int) bool { return bytes[a] == patChars[b] })
 }
 
-// DoMatch is a adapter for `DoMatchInner`, `str` can be any unicode string.
+// DoMatch is an adapter for `DoMatchInner`, `str` can be any unicode string.
 func DoMatch(str string, patChars []rune, patTypes []byte) bool {
-	return DoMatchInner(str, patChars, patTypes, matchRune)
```
Comment (YangKeao) [inline pkg/util/stringutil/string_util.go:238]: How about `return DoMatchCustomized(str, patChars, patTypes, matchRune)`.

**Review by YangKeao — COMMENTED** [review]
Rest LGTM.

BTW, I have checked that the behavior of TiKV is correct 👍 . It decodes the character according to the charset.

File: pkg/util/collate/bin.go (hunk 1 above)
Comment (yibin87) [reply pkg/util/collate/bin.go:69]: According to https://github.com/pingcap/tidb/pull/20610, some utf_bin collation needs to use Rune characters.

File: pkg/util/stringutil/string_util.go (hunk 2 above)
Comment (yibin87) [reply pkg/util/stringutil/string_util.go:285]: Make sense, I'll try.

File: pkg/util/collate/bin.go (hunk 1 above)
Comment (YangKeao) [reply pkg/util/collate/bin.go:69]: You are right. I realized that `utf8_bin` and `utf8mb4_bin` are using `binPaddingCollator`, and `utf8mb4_0900_bin` is using `binCollator` 🤦 (and also `gbkBinCollator` uses the `binPattern` directly).

**Comment by windtalker** [comment]
And please double check if TiFlash has the same issue

**Review by xiongjiwei — APPROVED** [review]

**Review by xzhangxian1008 — APPROVED** [review]
LGTM

File: pkg/util/stringutil/string_util.go [hunk 4]
```diff
@@ -228,53 +281,62 @@ func CompileLike2Regexp(str string) string {
...
-	for pIdx < len(patWeights) || rIdx < lenRunes {
-		if pIdx < len(patWeights) {
+	lenPatWeights := len(patWeights)
+	return DoMatchInner(lenPatWeights, lenRunes, patTypes, func(a, b int) bool { return matcher(runes[a], patWeights[b]) })
+}
+
+// DoMatchInner matches the string with patChars and patTypes.
+// The algorithm has linear time complexity.
+// https://research.swtch.com/glob
+func DoMatchInner(lenPatWeights int, lenChars int, patTypes []byte, matcher func(a, b int) bool) bool {
```
[the remark this answers is not in the document]
Comment (yibin87) [reply pkg/util/stringutil/string_util.go]: ok

File: pkg/util/stringutil/string_util.go [hunk 5]
```diff
@@ -138,21 +138,27 @@ const (
...
 func CompilePatternBytes(pattern string, escape byte) (patChars, patTypes []byte) {
 	patWeights, patTypes := CompilePatternInner(pattern, escape)
 	patChars = []byte(string(patWeights))
 
 	return patChars, patTypes
 }
 
-// CompilePattern is a adapter for `CompilePatternInner`, `pattern` can be any unicode string.
+// CompilePatternPureBytes is used for binary strings.
+func CompilePatternPureBytes(pattern string, escape byte) (patChars, patTypes []byte) {
```
[the remark this answers is not in the document]
Comment (yibin87) [reply pkg/util/stringutil/string_util.go]: ok

File: pkg/util/collate/collate.go [hunk 6]
```diff
@@ -400,8 +400,8 @@ func init() {
 	newCollatorMap = make(map[string]Collator)
 	newCollatorIDMap = make(map[int]Collator)
 
-	newCollatorMap["binary"] = &binCollator{}
-	newCollatorIDMap[CollationName2ID("binary")] = &binCollator{}
+	newCollatorMap["binary"] = &binPureCollator{}
```
[the remark this answers is not in the document]
Comment (yibin87) [reply pkg/util/collate/collate.go]: ok

**Comment by yibin87** [comment]
> And please double check if TiFlash has the same issue

The same issue detected in TiFlash, filed a issue to track it: https://github.com/pingcap/tiflash/issues/8776

**Review by windtalker — APPROVED** [review]
LGTM

**Pull request merged**